rclcpp_generic/include/rclcpp_generic/generic_publisher.hpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_GENERIC__GENERIC_PUBLISHER_HPP_
+#define RCLCPP_GENERIC__GENERIC_PUBLISHER_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/publisher_base.hpp"
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+#include "rclcpp/node_interfaces/node_topics_interface.hpp"
+#include "rclcpp/macros.hpp"
+#include "rclcpp/qos.hpp"
+#include "rclcpp/callback_group.hpp"
+#include "rcpputils/shared_library.hpp"
```

Comment:
Also, many of these could be forward declarations. ROS2's style guide (= Google style guide) says to avoid them, so I did, but they are present in many other packages.

Reply:
Forward declarations are usually only used in rclcpp (as far as I understand it) to resolve potentially circular includes - not just for convenience. It's definitely nice to know in the header that the definition is coming from _somewhere_